src/ttsim/interface_dag_elements/fail_if.py

```diff
@@ -304,15 +304,12 @@ def input_data_is_invalid(input_data__flat: FlatData) -> None:
         raise ValueError(msg)
 
     # Check for non-unique p_ids
-    p_id_counts: dict[int, int] = {}
-    # Need the map because Jax loop items are 1-element arrays.
-    for i in map(int, p_id):
-        if i in p_id_counts:
-            p_id_counts[i] += 1
-        else:
-            p_id_counts[i] = 1
+    # Convert to numpy array for performance (avoid slow JAX array iteration)
 
-    non_unique_p_ids = [i for i, count in p_id_counts.items() if count > 1]
+    # Use pandas duplicated for efficient duplicate detection
+    p_id_series = pd.Series(numpy.asarray(p_id))
+    duplicated_mask = p_id_series.duplicated(keep=False)
+    non_unique_p_ids = p_id_series[duplicated_mask].unique().tolist()
```

Comment:
I think we can keep this in numpy, it should be similarly fast. I noticed that we are sorting the p_id array quite often, maybe we should at some point combine all these operations, so that we only sort once.

```suggestion
    p_id_sorted = numpy.sort(p_id)
    duplicates = numpy.where(numpy.diff(p_id_sorted, append=0) == 0, True, False)

    if numpy.sum(duplicates) >= 1:
        message = (
            "The following `p_id`s are not unique in the input data:\n\n"
            f"{p_id_sorted[duplicates]}\n\n"
        )
        raise ValueError(message)
```

Reply:
### With NumPy instead of Pandas:
```
================================================================================
Households  NumPy (s)   JAX (s)     NumPy Hash  JAX Hash    Speedup
--------------------------------------------------------------------------------
32,767      0.0860      0.8723      2584ff76    9a97a0ef    1/10.15x
32,768      0.0693      0.5594      a4ec6c3f    c0a7c7e9    1/8.08x
65,536      0.1166      0.6014      e18ee207    f873c7c8    1/5.16x
131,072     0.2135      0.7101      e949fd41    84608262    1/3.33x
262,144     0.4691      1.1576      a8fe5730    2fb136ba    1/2.47x
524,288     0.9420      2.0237      c7c95469    dd61a524    1/2.15x
1,048,576   1.9857      3.7792      1fbbd4ff    6ed4af0a    1/1.90x
2,097,152   4.8578      8.1802      81209a75    ff9f600f    1/1.68x
```

Exactly as fast as Pandas.

### With JAX instead of NumPy

I used this code:
```python
    # Check for non-unique p_ids
    p_id_sorted = xnp.sort(p_id)
    duplicates = xnp.where(xnp.diff(p_id_sorted, append=0) == 0, True, False)
    if xnp.sum(duplicates) >= 1:
        message = (
            "The following `p_id`s are not unique in the input data:\n\n"
            f"{p_id_sorted[duplicates]}\n\n"
        )
        raise ValueError(message)
```

```
================================================================================
PERFORMANCE COMPARISON
================================================================================
Households  NumPy (s)   JAX (s)     NumPy Hash  JAX Hash    Speedup
--------------------------------------------------------------------------------
32,767      0.0708      0.8176      2584ff76    9a97a0ef    1/11.54x
32,768      0.0659      0.6828      a4ec6c3f    c0a7c7e9    1/10.37x
65,536      0.1160      0.7421      e18ee207    f873c7c8    1/6.40x
131,072     0.2134      0.8519      e949fd41    84608262    1/3.99x
262,144     0.4294      1.1789      a8fe5730    2fb136ba    1/2.75x
524,288     0.8776      2.0219      c7c95469    dd61a524    1/2.30x
1,048,576   1.9762      3.8733      1fbbd4ff    6ed4af0a    1/1.96x
2,097,152   4.5925      7.9699      81209a75    ff9f600f    1/1.74x
```

Too close to call...

However, I just noticed that while all three approaches pass all tests in TTSIM, only the currently commited approach passes all GETTSIM tests. I get a lot of `ValueError: The following `p_id`s are not unique in the input data` errors in GETTSIM. It seems the GETTSIM tests cover some corner cases that the TTSIM-Tests do not?

Reply:
Great analysis, thanks! 

Could you please
1. Add a test to TTSIM that makes the same error appear as in the GETTSIM test suite
2. Use the `xnp`-based implementation, adding an early return as below upfront?

```py
if p_id.shape[0] == 1:
    return
```

Reason is that before tackling that issue more generally (see comment above), I'd prefer to stick as close as possible to what we have rather than introducing a new strategy.